src/expr/src/scalar/like_pattern.rs

```diff
@@ -73,3 +158,774 @@ pub fn build_regex(pattern: &str, flags: &str) -> Result<Regex, EvalError> {
         ))),
     }
 }
+
+// The algorithm below is based on the observation that any LIKE pattern can be
+// decomposed into multiple parts:
+//     <PATTERN> := <SUB-PATTERN> (<SUB-PATTERN> ...)
+//     <SUB-PATTERN> := <WILDCARDS> <SUFFIX>
+//
+// The sub-patterns start with zero or more wildcard characters, eventually
+// followed by (non-wildcard) literal characters. The last sub-pattern may
+// have an empty SUFFIX.
+//
+// Example: the PATTERN "n__dl%" can be broken into the following parts:
+//   1. SUB-PATTERN = <WILDCARDS ""> <SUFFIX "n">
+//   2. SUB-PATTERN = <WILDCARDS "__"> <SUFFIX "dl">
+//   3. SUB-PATTERN = <WILDCARDS "%"> <SUFFIX "">
+//
+// The WILDCARDS can be any combination of '_', which matches exactly 1 char,
+// and '%' which matches zero or more chars. These wildcards can be simplified
+// down to the (min, max) of characters they might consume:
+//     ""  = (0, 0)    // doesn't consume any characters
+//     "_" = (1, 1)    // consumes exactly one
+//     "%" = (0, many) // zero or more
+// These are additive, so:
+//     "_%"    = (1, many)
+//     "__%__" = (4, many)
+//     "%%%_"  = (1, many)
+
+// Matches a chain of SUB-PATTERNs
+#[derive(Debug, Clone, Deserialize, Serialize)]
+struct StringMatcher {
+    /// The minimum number of characters that can be consumed by the wildcard expression.
+    consume: usize,
+    /// Whether the wildcard expression can consume an arbitrary number of characters.
+    many: bool,
+    /// A string literal that is expected after the wildcards.
+    suffix: String,
```

Comment:
A future optimization could swap the `String` for indexes into the pattern string (which you're keeping around in the `Matcher` anyway) to avoid allocating, but that's probably not worth the extra complexity at the moment.

Reply:
Ah, right, exactly, that's why you'd want to store indexes into the pattern string rather than `&str` references.

Quick aside: if you're lucky, you can design the API so that the caller is responsible for storing the string:

```rust
struct CompiledMatcher<'a> {
    subpatterns: Vec<&'a str>
}

fn compile_matcher<'a>(s: &'a str) -> CompiledMatcher<'a> {
    // You can construct and return slices to `s` freely.
}

fn check_for_matches(pattern: &str, inputs: &[str]) -> bool {
    // Rust can infer that the returned matcher does not outlive `pattern`.
    let matcher = compile_matcher(pattern);
    inputs.iter().any(|inp| matcher.match(inp))
}
```

That trick won't work here because we want to store the pattern string in the matcher struct itself. (You can't have a self-referential struct in Rust without `unsafe` code.)

The other trick is to avoid Rust references and instead use a `Range<usize>` that describes the start and end indices in the string. That avoids lifetime issues by avoiding the pointer. Memory-wise it's exactly as efficient (a slice is 64-bits); compute-wise I think there's one extra bounds check required on the start index, but pretty rare for that to matter.